File: matlab/ode_modelling/Scaling.m
classdef Scaling
    properties(SetAccess=private)
        factor
        offset
    end

    methods
        function obj = Scaling(factor, offset)
            obj.factor = factor;

            if (nargin == 1)
                obj.offset = 0;
            else
                obj.offset = offset;
            end
        end

        function unscale = get_unscale_scaling(obj)
            unscale = Scaling(1 / obj.factor, -obj.offset / obj.factor);
        end

        function value = scale(obj, value)
            value = (value - obj.offset) / obj.factor;
        end

        function value = unscale(obj, value)
            value = value * obj.factor + obj.offset;
        end

        function value = scale_derivate(obj, value)
            value = value / obj.factor;
        end

        function value = unscale_derivate(obj, value)
            value = value * obj.factor;
        end
    end

    methods(Static)
        function scaling = FromRange(min, max)
            factor = (max - min) / 2;
            offset = (max + min) / 2;

            scaling = Scaling(factor, offset);
        end
    end
end

File: python/src/ctrldesign/ode_modelling/scaling.py
from dataclasses import dataclass


@dataclass
class Scaling:
    factor: float
    offset: float

    def get_unscale_scaling(self):
        return Scaling(1 / self.factor, -self.offset / self.factor)

    def scale(self, value):
        return (value - self.offset) / self.factor

    def unscale(self, value):
        return value * self.factor + self.offset

    def scale_derivate(self, value):
        return value / self.factor

    def unscale_derivate(self, value):
        return value * self.factor

    @staticmethod
    def from_range(min: float, max: float):
        factor = (max - min) / 2
        offset = (max + min) / 2

        return Scaling(factor, offset)
